### DISCUSSION TITLE
_WidgetHandleImpl

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#145] [General]
Can you explain this implementation (and why exists the `if` clause)?

```dart
@override
  T register<T>(WidgetSideEffect<T> sideEffect) {
    [...]

    if (sideEffectDataIndex == api.manager.sideEffectData.length) {
      api.manager.sideEffectData.add(sideEffect(api));
    }
    return api.manager.sideEffectData[sideEffectDataIndex++] as T;
  }
```

**Comment by GregoryConrad** [comment]
On the first build, the array will be empty so that checks if it’s a new side effect that’s being called, and if so, adds a new entry in the list for it. On subsequent builds, it just returns the side effect at that particular index

**Reply by busslina** [reply]
So here you could check, on not the first build, a new side effect adding attempt (bad usage) and throw an error.
But yeah, I understand you don't do that for now:
- Performance reason
- You might also want to check not just side effect index (or count) but type too
- ...

**Reply by GregoryConrad** [reply]
It’d be easy enough to add that kind of assert in; I just never bothered haha
